pkg/trace/api/normalizer_test.go

```diff
@@ -33,277 +36,326 @@ func testSpan() *pb.Span {
 	}
 }
 
+// assertNormalizationIssue asserts there is exactly one NormalizationIssue in the provided TagStats matching the reason
+func assertNormalizationIssue(t *testing.T, ts *info.TagStats, reason string) {
+	allIssues := ts.AllNormalizationIssues()
+
+	presentIssues := make(map[string]int64)
+	for r, c := range allIssues {
+		if c > 0 {
+			presentIssues[r] = c
+		}
+	}
+
+	foundIssuesMessage := "none"
+	if len(presentIssues) > 0 {
+		s, _ := json.Marshal(presentIssues)
+		foundIssuesMessage = string(s)
+	}
+
+	if count, ok := presentIssues[reason]; ok {
+		assert.EqualValues(t, 1, count, fmt.Sprintf("expected normalization issue %s=%d. Found issues: %s", reason, 1, foundIssuesMessage))
+	} else {
+		assert.Fail(t, fmt.Sprintf("missing expected normalization issue %s=%d. Found issues: %s", reason, 1, foundIssuesMessage))
+	}
+}
+
+// assertNormalizationIssue asserts there there are no NormalizationIssues in the provided TagStats
+func assertNoNormalizationIssues(t *testing.T, ts *info.TagStats) {
+	for reason, count := range ts.AllNormalizationIssues() {
+		assert.EqualValues(t, 0, count, fmt.Sprintf("found unexpected normalization issue %s=%d, expecting no issues", reason, count))
+	}
+}
+
 func TestNormalizeOK(t *testing.T) {
-	s := testSpan()
-	assert.NoError(t, normalize(s))
+	ts, s := &info.TagStats{}, newTestSpan()
+	assert.NoError(t, normalize(ts, s))
+	assertNoNormalizationIssues(t, ts)
```

Comment:
Yes, but if there is an unexpected normalization issue then the log message won't be as descriptive. As it is now it will say `found unexpected normalization issue %s=%d`, with the reason matching the reason names used everywhere else. So I think better leave it like that. And it's only a couple of lines of code for `assertNoNormalizationIssues` so it's not like switching to that would make things significantly simpler.

Reply:
Updated as explained in https://github.com/DataDog/datadog-agent/pull/3738#discussion_r297788738